src/plugins/camera/camera_impl.cpp

```diff
@@ -1583,4 +1612,117 @@ void CameraImpl::format_storage_async(Camera::ResultCallback callback)
         cmd_format, std::bind(&CameraImpl::receive_command_result, this, _1, callback));
 }
 
+std::pair<Camera::Result, std::vector<Camera::CaptureInfo>>
+CameraImpl::list_photos(Camera::PhotosRange photos_range)
+{
+    auto prom = std::make_shared<
+        std::promise<std::pair<Camera::Result, std::vector<Camera::CaptureInfo>>>>();
+    auto ret = prom->get_future();
+
+    list_photos_async(
+        photos_range, [prom](Camera::Result result, std::vector<Camera::CaptureInfo> photo_list) {
+            prom->set_value(std::make_pair(result, photo_list));
+        });
+
+    return ret.get();
+}
+
+void CameraImpl::list_photos_async(
+    Camera::PhotosRange photos_range, const Camera::ListPhotosCallback callback)
+{
+    {
+        std::lock_guard<std::mutex> lock(_status.mutex);
+
+        if (_status.is_fetching_photos) {
+            _parent->call_user_callback([callback]() {
+                callback(Camera::Result::Busy, std::vector<Camera::CaptureInfo>{});
+            });
+            return;
+        } else {
+            _status.is_fetching_photos = true;
+        }
+
+        if (_status.image_count == -1) {
+            LogErr() << "Cannot list photos: camera status has not been received yet!";
+            _parent->call_user_callback([callback]() {
+                callback(Camera::Result::Error, std::vector<Camera::CaptureInfo>{});
+            });
+            return;
+        }
+    }
+
+    int start_index;
+
+    switch (photos_range) {
+        case Camera::PhotosRange::SinceConnection:
+            start_index = _status.image_count_at_connection;
+            break;
+        case Camera::PhotosRange::All:
+        default:
+            start_index = 0;
+            break;
+    }
+
+    std::thread([this, start_index, callback]() {
```

Comment:
Have you considered doing this with `call_every` instead of a thread?

Reply:
I was not comfortable waiting on a condition variable in `call_every`. That would block the event loop, wouldn't it?

Reply:
Nevermind, leave it as is!